### PR TITLE
Bug 1474570 - Migrate linux talos workers from task cluster to generic worker

### PR DESCRIPTION
**Pull request by dragoscrisan**
Added Ubuntu 16 capabilities to generic-worker module.
Modified generic-worker to automatically create configuration file for Linux and OSX workers
Added t-linux64-ms-280.test.releng.mdc1.mozilla.com to staging pool
Enabled http_proxy module

### CODE REVIEW

File: modules/generic_worker/manifests/init.pp [hunk 1]
```diff
@@ -66,6 +79,72 @@
                     content => template('generic_worker/proxy-httpd.conf.erb');
             }
         }
+        Ubuntu: {
+            case $::operatingsystemrelease {
+                16.04: {
+                    if ($environment == 'staging') {
+                        # We are limited to 22 characters for worker_type
+                        $worker_type = "gecko-t-linux-talos-b"
```
Comment (dividehex) [inline modules/generic_worker/manifests/init.pp]: swap double quotes for single

File: modules/generic_worker/manifests/init.pp [hunk 2]
```diff
@@ -66,6 +79,72 @@
...
+                        $taskcluster_client_id = secret('generic_worker_linux_staging_client_id')
+                        $taskcluster_access_token = hiera('generic_worker_linux_staging_access_token')
+                    }
+                    else {
+                        $worker_type = "gecko-t-osx-linux-talos"
```
Comment (dividehex) [inline modules/generic_worker/manifests/init.pp]: swap double quotes for single

File: modules/generic_worker/manifests/init.pp [hunk 3]
```diff
@@ -66,6 +79,72 @@
...
+                        group   => $users::root::group;
+                    }
+                    file { '/etc/generic-worker.config':
+                        ensure  => present,
+                        content => template('generic_worker/generic-worker.config.erb'),
+                        mode    => '0644',
+                        owner   => $users::root::username,
+                        group   => $users::root::group;
+                    }
+                    exec { 'create gpg key':
```
Comment (dividehex) [inline modules/generic_worker/manifests/init.pp]: This exec looks like it is already defined on line 66.  You probably don't want to duplicate that.

File: modules/generic_worker/manifests/init.pp [hunk 4]
```diff
@@ -66,6 +79,72 @@
...
+                        user    => 'cltbld',
+                        cwd     => $users::builder::home,
+                        command => "generic-worker new-openpgp-keypair --file ${::users::builder::home}/generic-worker.openpgp.key",
+                        unless  => "test -f ${::users::builder::home}/generic-worker.openpgp.key"
+                    }
+
+                    host {"${taskcluster_host}":
+                        ip => '127.0.0.1'
+                    }
+                    # Enable proxy_http_module on apache2
```
Comment (dividehex) [inline modules/generic_worker/manifests/init.pp]: Make sure you are explicitly including the httpd (apache) module if you are going to be calling a2enmod and notifying the service defined there.  You might even want to require it just to be on the safe side.

File: modules/generic_worker/templates/generic-worker.config.erb [hunk 5]
```diff
@@ -1,19 +1,19 @@
 {
   "accessToken": "<%= @taskcluster_access_token %>",
-  "cachesDir": "/Users/cltbld/caches",
+  "cachesDir": "<%= @caches_dir %>",
```
Comment (dividehex) [inline modules/generic_worker/templates/generic-worker.config.erb]: 👍 to sourcing these from manifest variables

File: modules/generic_worker/templates/gnome-terminal.desktop.erb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+[Desktop Entry]
+Type=Application
+Exec=gnome-terminal -x /usr/local/bin/run-generic-worker.sh run --config /etc/generic-worker.config
+Hidden=false
+X-GNOME-Autostart-enabled=true
+Name=Generic-Worker
+Comment=Start generic-worker in a terminal session
+StartupNotify=false
+Terminal=false
+Type=Application
```
Comment (dividehex) [inline modules/generic_worker/templates/gnome-terminal.desktop.erb]: Make sure to drop a newline at the end of this file

**Comment by dragoscrisan** [comment]
I added the hosts definition for staging workers into this PR because I want to avoid conflicts when we will merge this PR

File: manifests/moco-nodes.pp [hunk 7]
```diff
@@ -1177,48 +1177,42 @@
...
+node 't-linux64-ms-280.test.releng.mdc1.mozilla.com',
+    't-linux64-ms-240.test.releng.mdc1.mozilla.com',
+    't-linux64-ms-394.test.releng.mdc2.mozilla.com',
+    't-linux64-ms-395.test.releng.mdc2.mozilla.com' {
+    $aspects = [ 'low-security' ]
+    $slave_trustlevel  = 'try'
+    $tc_environment       = 'staging'
+    include fw::profiles::osx_taskcluster_worker
+    include toplevel::worker::releng::generic_worker::test::gpu
+}
```
Comment (dividehex) [inline manifests/moco-nodes.pp]: Add newline before EOF

File: modules/generic_worker/manifests/init.pp [hunk 8]
```diff
@@ -118,22 +121,14 @@
...
                         ip => '127.0.0.1'
                     }
                     # Enable proxy_http_module on apache2
                     exec { 'enable proxy_http_module':
-                        path    => ['/bin', '/sbin', '/usr/local/bin', '/usr/bin', '/usr/sbin'],
-                        command => 'a2enmod proxy_http',
-                        notify  => Service['httpd'];
+                        path      => ['/bin', '/sbin', '/usr/local/bin', '/usr/bin', '/usr/sbin'],
+                        command   => 'a2enmod proxy_http',
+                        subscribe => Service['httpd'];
```
Comment (dividehex) [inline modules/generic_worker/manifests/init.pp]: I think 'notify' was the correct relationship metaparameter to set here.  Since a2enmod sets up symbolic links in the config directories, I believe you will want the service to be restarted _**after**_ the a2enmod has executed.  My previous comments might have been a little confusing.  Sorry about that.

You should also have an 'onlyif, unless, or creates' to prevent this from executing on every run.  I would suggest using 'creates' to check for the symlink in /etc/apache2/mods-enabled

http://manpages.ubuntu.com/manpages/bionic/man8/a2enmod.8.html

**Comment by davehouse** [comment]
+1 (re-up my review)

File: manifests/moco-nodes.pp [hunk 9]
```diff
@@ -1177,46 +1177,42 @@
...
-node 't-yosemite-r7-380.test.releng.mdc1.mozilla.com' {
+# Workers in osx staging pool
+node 't-yosemite-r7-380.test.releng.mdc1.mozilla.com',
+    't-yosemite-r7-394.test.releng.mdc1.mozilla.com',
+    't-yosemite-r7-100.test.releng.mdc2.mozilla.com',
+    't-yosemite-r7-101.test.releng.mdc2.mozilla.com' {
     $aspects           = [ 'low-security' ]
     $slave_trustlevel  = 'try'
-    $environment       = 'staging'
+    $tc_environment       = 'staging'
```
Comment (petemoore) [inline manifests/moco-nodes.pp]: I'm curious: why the change from `$environment` to `$tc_environment`?

File: manifests/moco-nodes.pp [hunk 10]
```diff
@@ -1177,46 +1177,42 @@
...
+
+
+# Workers in linux talos staging pool
+node 't-linux64-ms-280.test.releng.mdc1.mozilla.com',
+    't-linux64-ms-240.test.releng.mdc1.mozilla.com',
+    't-linux64-ms-394.test.releng.mdc2.mozilla.com',
+    't-linux64-ms-395.test.releng.mdc2.mozilla.com' {
+    $aspects = [ 'low-security' ]
+    $slave_trustlevel  = 'try'
+    $tc_environment       = 'staging'
```
Comment (petemoore) [inline manifests/moco-nodes.pp]: nit: spacing is off here

File: modules/generic_worker/templates/run-generic-worker.sh.erb [hunk 11]
```diff
@@ -49,14 +49,20 @@ if [ $exitstatus -eq 69 ]; then
...
-    /usr/bin/sudo /sbin/reboot
-    # Sleep to prevent this script from terminating naturally, and launchd restarting
-    # it. Instead, the shutdown should cause this script to terminate (so it won't
-    # really sleep for 2 mins). If shutdown doesn't kick in within 2 mins, it is sane
-    # for this script to exit, and allow launchd to fire up the worker again.
-    /bin/sleep 120
+    # Remove quarantined file, if file exist
+    if [ -f <%= scope.lookupvar('users::builder::home') %>/quarantined ]; then
+        /bin/rm -rf <%= scope.lookupvar('users::builder::home') %>/quarantined
+    fi
```
Comment (petemoore) [inline modules/generic_worker/templates/run-generic-worker.sh.erb]: The `-f` option already should only remove the file if it exists. The `-r` is only needed for recursive deletes of subdirectories - are there subdirectories that also get deleted?

**Comment by davehouse** [comment]
+1 updating my review again. looks good
